File: extensions/vscode/src/contextProvider.ts
/**
 * IDE context provider — collects active file, selection, and diagnostics
 * from VS Code APIs and sends them to the Claude Code process.
 */

import * as vscode from 'vscode'
import type { ProcessManager } from './processManager'
import type {
  IDEFileContext,
  IDESelection,
  IDEDiagnostic,
  IDEContextMessage,
} from './protocol'

export class ContextProvider implements vscode.Disposable {
  private disposables: vscode.Disposable[] = []
  private debounceTimer: ReturnType<typeof setTimeout> | null = null
  private debounceMs = 500

  constructor(
    private processManager: ProcessManager,
    private enabled: () => boolean,
    private onContextSent?: (context: IDEContextMessage) => void,
  ) {
    this.setupListeners()
  }

  private setupListeners(): void {
    // Active editor changes
    this.disposables.push(
      vscode.window.onDidChangeActiveTextEditor(() => {
        this.scheduleContextUpdate()
      }),
    )

    // Selection changes
    this.disposables.push(
      vscode.window.onDidChangeTextEditorSelection(() => {
        this.scheduleContextUpdate()
      }),
    )

    // Document changes (for diagnostics refresh)
    this.disposables.push(
      vscode.workspace.onDidChangeTextDocument(() => {
        // Only update context for visible editors to avoid noise
        if (vscode.window.activeTextEditor) {
          this.scheduleContextUpdate()
        }
      }),
    )

    // Diagnostics changes
    this.disposables.push(
      vscode.languages.onDidChangeDiagnostics(() => {
        this.scheduleContextUpdate()
      }),
    )
  }

  /** Send current context immediately */
  async sendContext(): Promise<void> {
    const context = await this.collectContext()
    if (context) {
      this.processManager.send(context)
      this.onContextSent?.(context)
    }
  }

  dispose(): void {
    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer)
    }
    for (const d of this.disposables) {
      d.dispose()
    }
  }

  // ====================================================================
  // Internal
  // ====================================================================

  private scheduleContextUpdate(): void {
    if (!this.enabled()) return

    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer)
    }
    this.debounceTimer = setTimeout(() => {
      this.sendContext()
    }, this.debounceMs)
  }

  private async collectContext(): Promise<IDEContextMessage | null> {
    const config = vscode.workspace.getConfiguration('claudeCode')
    const maxFiles = config.get<number>('maxFilesInContext', 5)

    const files = await this.collectActiveFiles(maxFiles)
    const selection = this.collectSelection()
    const diagnostics = this.collectDiagnostics()

    if (files.length === 0 && !selection && diagnostics.length === 0) {
      return null
    }

    return {
      type: 'ide_context',
      files: files.length > 0 ? files : undefined,
      selection,
      diagnostics: diagnostics.length > 0 ? diagnostics : undefined,
    }
  }

  private async collectActiveFiles(maxFiles: number): Promise<IDEFileContext[]> {
    const result: IDEFileContext[] = []
    const seen = new Set<string>()

    // Add active editor first
    const activeEditor = vscode.window.activeTextEditor
    if (activeEditor) {
      const doc = activeEditor.document
      seen.add(doc.uri.fsPath)
      result.push({
        path: vscode.workspace.asRelativePath(doc.uri),
        content: doc.getText(),
        language: doc.languageId,
      })
    }

    // Add visible editors
    for (const editor of vscode.window.visibleTextEditors) {
      if (result.length >= maxFiles) break
      const doc = editor.document
      if (!seen.has(doc.uri.fsPath)) {
        seen.add(doc.uri.fsPath)
        result.push({
          path: vscode.workspace.asRelativePath(doc.uri),
          content: doc.getText(),
          language: doc.languageId,
        })
      }
    }

    return result
  }

  private collectSelection(): IDESelection | null {
    const editor = vscode.window.activeTextEditor
    if (!editor || editor.selection.isEmpty) return null

    const sel = editor.selection
    const doc = editor.document

    return {
      path: vscode.workspace.asRelativePath(doc.uri),
      startLine: sel.start.line,
      startChar: sel.start.character,
      endLine: sel.end.line,
      endChar: sel.end.character,
      text: doc.getText(sel),
    }
  }

  private collectDiagnostics(): IDEDiagnostic[] {
    const result: IDEDiagnostic[] = []
    const activeEditor = vscode.window.activeTextEditor
    if (!activeEditor) return result

    const activeUri = activeEditor.document.uri
    const diags = vscode.languages.getDiagnostics(activeUri)

    for (const d of diags.slice(0, 20)) {
      // Map VS Code severity
      let severity: IDEDiagnostic['severity']
      switch (d.severity) {
        case vscode.DiagnosticSeverity.Error:
          severity = 'error'
          break
        case vscode.DiagnosticSeverity.Warning:
          severity = 'warning'
          break
        case vscode.DiagnosticSeverity.Information:
          severity = 'info'
          break
        case vscode.DiagnosticSeverity.Hint:
          severity = 'hint'
          break
        default:
          severity = 'info'
      }

      result.push({
        path: vscode.workspace.asRelativePath(activeUri),
        line: d.range.start.line,
        column: d.range.start.character,
        message: d.message,
        severity,
        source: d.source,
      })
    }

    return result
  }
}

File: extensions/vs/ClaudeCodeVS/Services/ContextProvider.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.Composition;
using System.IO;
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;
using System.Threading;
using ClaudeCodeVS.Messages;

namespace ClaudeCodeVS.Services
{
    /// <summary>
    /// Collects IDE context (active document, selection, diagnostics) and
    /// sends it to the Bun backend via WebSocket.
    ///
    /// Ported from extensions/vscode/src/contextProvider.ts.
    /// VS equivalents:
    ///   vscode.window.activeTextEditor     → IWpfTextView (via MEF listener)
    ///   vscode.window.visibleTextEditors   → Enumerate open ITextDocument
    ///   vscode.workspace.asRelativePath    → Path.GetRelativePath(solutionDir, filePath)
    ///   vscode.languages.getDiagnostics    → Error List / IVsSolutionBuildManager
    /// </summary>
    public class ContextProvider : IDisposable
    {
        private readonly Func<bool> _enabled;
        private readonly Action<string> _sendToBackend;  // sends JSON to WebSocket
        private readonly CancellationTokenSource _cts = new CancellationTokenSource();

        private int _maxFilesInContext = 5;
        private System.Timers.Timer _debounceTimer;

        public ContextProvider(Func<bool> enabled, Action<string> sendToBackend)
        {
            _enabled = enabled;
            _sendToBackend = sendToBackend;
        }

        /// <summary>
        /// Call from IWpfTextViewConnectionListener when a text view is connected.
        /// This replaces vscode.window.onDidChangeActiveTextEditor.
        /// </summary>
        public void OnTextViewActivated(Microsoft.VisualStudio.Text.Editor.IWpfTextView textView)
        {
            var doc = GetTextDocument(textView);
            if (doc == null) return;

            ScheduleContextUpdate();
        }

        /// <summary>
        /// Call from ITextSelection.SelectionChanged.
        /// </summary>
        public void OnSelectionChanged()
        {
            ScheduleContextUpdate();
        }

        /// <summary>
        /// Call from ITextBuffer.Changed on visible documents.
        /// </summary>
        public void OnDocumentChanged()
        {
            ScheduleContextUpdate();
        }

        /// <summary>
        /// Call from a diagnostics change event (Error List or Roslyn listener).
        /// </summary>
        public void OnDiagnosticsChanged()
        {
            ScheduleContextUpdate();
        }

        // ── Context Collection ─────────────────────────────────────────────

        public void SendContext()
        {
            if (_sendToBackend == null) return;

            var context = CollectContext();
            if (context == null) return;

            var json = JsonConvert.SerializeObject(context, new JsonSerializerSettings
            {
                ContractResolver = new CamelCasePropertyNamesContractResolver(),
                NullValueHandling = NullValueHandling.Ignore,
            });
            _sendToBackend(json);
        }

        private IdeContextMessage CollectContext()
        {
            var files = CollectActiveFiles(_maxFilesInContext);
            var selection = CollectSelection();
            var diagnostics = CollectDiagnostics();

            if (files.Count == 0 && selection == null && diagnostics.Count == 0)
                return null;

            return new IdeContextMessage
            {
                Type = MessageTypes.IdeContext,
                Files = files.Count > 0 ? files : null,
                Selection = selection,
                Diagnostics = diagnostics.Count > 0 ? diagnostics : null,
            };
        }

        private List<FileContext> CollectActiveFiles(int maxFiles)
        {
            var result = new List<FileContext>();

            try
            {
                // Use DTE to enumerate open documents
                // var dte = (EnvDTE80.DTE2)ServiceProvider.GetService(typeof(EnvDTE.DTE));
                var dte = GetDTE();
                if (dte == null) return result;

                var activeDoc = dte.ActiveDocument;
                if (activeDoc != null)
                {
                    var content = GetDocumentContent(activeDoc.FullName);
                    if (content != null)
                    {
                        result.Add(new FileContext
                        {
                            Path = activeDoc.FullName,
                            Content = content,
                            Language = activeDoc.Language,
                        });
                    }
                }

                // TODO: Add visible tab documents (not just active)
                // For now, active document is the primary context source
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"[ClaudeCodeVS] Context error: {ex.Message}");
            }

            return result;
        }

        private Selection CollectSelection()
        {
            try
            {
                var dte = GetDTE();
                if (dte == null) return null;

                var textSelection = dte.ActiveDocument?.Selection as EnvDTE.TextSelection;
                if (textSelection == null || textSelection.IsEmpty) return null;

                var text = textSelection.Text;
                if (string.IsNullOrEmpty(text)) return null;

                return new Selection
                {
                    Path = dte.ActiveDocument.FullName,
                    StartLine = textSelection.TopLine - 1,       // 1-based → 0-based
                    StartChar = textSelection.TopPoint.LineCharOffset - 1,
                    EndLine = textSelection.BottomLine - 1,
                    EndChar = textSelection.BottomPoint.LineCharOffset - 1,
                    Text = text,
                };
            }
            catch
            {
                return null;
            }
        }

        private List<Messages.Diagnostic> CollectDiagnostics()
        {
            // Simplified: VS diagnostics require Error List Provider integration.
            // For initial release, return empty. Can be enhanced later.
            return new List<Messages.Diagnostic>();
        }

        // ── Helper Methods ────────────────────────────────────────────────

        private void ScheduleContextUpdate()
        {
            if (!_enabled()) return;

            _debounceTimer?.Stop();
            _debounceTimer = new System.Timers.Timer(500) { AutoReset = false };
            _debounceTimer.Elapsed += (s, e) =>
            {
                SendContext();
            };
            _debounceTimer.Start();
        }

        private static EnvDTE80.DTE2 GetDTE()
        {
            try
            {
                return Microsoft.VisualStudio.Shell.Package.GetGlobalService(
                    typeof(EnvDTE.DTE)) as EnvDTE80.DTE2;
            }
            catch
            {
                return null;
            }
        }

        private static string GetDocumentContent(string filePath)
        {
            try
            {
                if (File.Exists(filePath))
                {
                    var content = File.ReadAllText(filePath);
                    // Cap at 100KB to avoid huge payloads
                    if (content.Length > 100 * 1024)
                    {
                        content = content.Substring(0, 100 * 1024) + "\n... (truncated)";
                    }
                    return content;
                }
            }
            catch { }
            return null;
        }

        private static Microsoft.VisualStudio.Text.ITextDocument GetTextDocument(
            Microsoft.VisualStudio.Text.Editor.IWpfTextView textView)
        {
            textView.TextBuffer.Properties.TryGetProperty(
                typeof(Microsoft.VisualStudio.Text.ITextDocument),
                out Microsoft.VisualStudio.Text.ITextDocument doc);
            return doc;
        }

        public void Dispose()
        {
            _cts.Cancel();
            _debounceTimer?.Dispose();
        }
    }
}
